### ISSUE TITLE
VS Code typescript language server failed to import types

### ISSUE THREAD

**Issue by AnsonYeung** [patrickjuchli/basic-ftp#49]
An notice will be shown when including the package.
```js
const ftp = require("basic-ftp");
```
The notice:
```
[ts]
Could not find a declaration file for module 'basic-ftp'. %MyFolder%/node_modules/basic-ftp/lib/ftp.js' implicitly has an 'any' type.
  Try `npm install @types/basic-ftp` if it exists or add a new declaration (.d.ts) file containing `declare module 'basic-ftp';` [7016]
```
Well, I'll say it doesn't really relate to the module.  
However, I would like to know whether any plan to get this working since many other npm package seems to work.  
After some investigation, I believe other npm modules works because someone contributes on [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) which publishes the npm package `@types/npm-package-name`.  
I think it is important becuase VS code is currently a very popular editor.

**Comment by patrickjuchli** [comment]
The source itself defines types using JSDoc sufficiently to satisfy even `strict` and `noImplicitAny` compiler settings. In VSCode, just using `checkJs` with a regular Javascript project works well.

But I haven't tested this with an actual Typescript project. I will have a look but I want to make it work using JSDoc only.

**Comment by AnsonYeung** [comment]
So my code is actually written in JavaScript. Any idea that I did something wrong? I use jsdoc in my code and it functions as expected, but it needs a declaration file for node module?

**Comment by AnsonYeung** [comment]
Let me provide more details.
```js
/**
 * @typedef {import("basic-ftp")} ftp
 */
```
And a message will be on `"basic-ftp"`
```
[ts]
Could not find a declaration file for module 'basic-ftp'. 'c:/Users/secur/Website/node_modules/basic-ftp/lib/ftp.js' implicitly has an 'any' type.
  Try `npm install @types/basic-ftp` if it exists or add a new declaration (.d.ts) file containing `declare module 'basic-ftp';` [7016]
```
However, when you hover over the type `ftp`, This will gives
```ts
type ftp = {
    Client: typeof Client;
    FTPContext: typeof FTPContext;
    FTPError: typeof FTPError;
    FileInfo: typeof FileInfo;
    utils: {
        upgradeSocket: (socket: Socket, options: ConnectionOptions) => Promise<TLSSocket>;
        parseIPv4PasvResponse: (message: string) => {
            ...;
        };
        parseIPv6PasvResponse: (message: string) => number;
    };
}
```
It doesn't seems to work when you declare the type, any of these will fail and fall back to `any` type.
```js
/**
 * @type {import("basic-ftp").Client}
 * @type {ftp.Client}
 * @type {Client}
 */
```
In my scenario, it was an empty array so I need to declare the type by JSDoc.
```js
/**
 * @type {ftp.Client[]}
 */
const myArray = [];
```
Do you have any idea why it doesn't work?  
While VSCode is able to show the type, it stills show an message. Should it be a issue of VSCode?

**Comment by patrickjuchli** [comment]
Using the following two lines with `"javascript.implicitProjectConfig.checkJs": true` works fine in VSCode: It shows an error for my first argument when trying to instantiate ftp.Client.

```
const ftp = require("basic-ftp")
const bla = new ftp.Client("VSCode SHOULD MARK ERROR HERE");
```

**Comment by patrickjuchli** [comment]
But it looks like there's something missing for Typescript projects. I will look into it.

**Comment by deziev** [comment]
@patrickjuchli hi, i think the best way to solve this issue would be adding native typings to repo. Im talking about `index.d.ts` file which defines package typings. That will significantly help to use basic-ftp in typescript projects. 
Anyway, i'd happy to make pull request with described feature.

**Comment by patrickjuchli** [comment]
There are some open issues on the Typescript repository about leveraging JSDoc annotations instead of having to specify type definitions a second time. It seems they want to support that. In the short term, we probably do need a `d.ts` type definition file. I don't like the idea of having two separate type definitions for the same things, though. Go ahead with the pull request, if you want to give it a try.

**Comment by patrickjuchli** [comment]
There is no elegant way to solve this except by rewriting the library in Typescript. Typescript doesn't let users use JSDoc when using Typescript. Conversion tools exist but they are either not maintained well or don't work well enough.

This is not as crazy as it sounds because everything is already strictly typed using JSDoc. Should be straightforward, let's go this way, work has already started.

**Comment by patrickjuchli** [comment]
Fixed in 3.2.0
